cpp/mrc/include/mrc/node/source_properties.hpp

```diff
@@ -208,37 +209,68 @@ template <typename T>
 class ForwardingEgressProvider : public ReadableProvider<T>
 {
   protected:
+    struct State
+    {
+        Mutex m_mutex;
+        bool m_is_destroyed{false};
+    };
+
     class ForwardingEdge : public edge::IEdgeReadable<T>
     {
       public:
-        ForwardingEdge(ForwardingEgressProvider<T>& parent) : m_parent(parent) {}
+        ForwardingEdge(ForwardingEgressProvider<T>& parent, std::shared_ptr<State> state) :
+          m_parent(parent),
+          m_state(std::move(state))
+        {}
 
         ~ForwardingEdge() = default;
 
         channel::Status await_read(T& t) override
         {
-            return m_parent.get_next(t);
+            std::lock_guard<decltype(m_state->m_mutex)> lock(m_state->m_mutex);
+            if (!(m_state->m_is_destroyed))
+            {
+                return m_parent.get_next(t);
+            }
+
+            return channel::Status::closed;
         }
 
       private:
         ForwardingEgressProvider<T>& m_parent;
+        std::shared_ptr<State> m_state;
     };
 
-    ForwardingEgressProvider()
+    ForwardingEgressProvider() : m_state(std::make_shared<State>())
     {
-        auto inner_edge = std::make_shared<ForwardingEdge>(*this);
+        auto inner_edge = std::make_shared<ForwardingEdge>(*this, m_state);
 
-        inner_edge->add_disconnector([this]() {
-            // Only call the on_complete if we have been connected
-            this->on_complete();
+        inner_edge->add_disconnector([this, state = m_state]() {
+            std::lock_guard<decltype(state->m_mutex)> lock(state->m_mutex);
+            if (!(state->m_is_destroyed))
+            {
+                // Only call the on_complete if we have been connected and `this` is still alive
+                this->on_complete();
```

Comment:
Need to see if there are any implementations of `on_complete`. If there arent, then we should just scrap the disconnector and remove the state and mutex changes.

Reply:
I checked, the classes which implement `on_complete` are:
* `mrc::node::NodeComponent` Calls `SourceProperties<OutputT>::release_edge_connection();` and `do_on_complete` which is unimpl in the base
* `mrc::node::RouterBase` calls `MultiSourceProperties<KeyT, output_data_t>::release_edge_connections();`

`do_on_complete` in turn is used by `LambdaNodeComponent` to call `m_on_complete_fn` if defined.